pkg/controllers/machine/disruption/drift.go

```diff
@@ -12,62 +12,58 @@ See the License for the specific language governing permissions and
 	"knative.dev/pkg/logging"
-	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 
 	"github.com/aws/karpenter-core/pkg/apis/settings"
 	"github.com/aws/karpenter-core/pkg/apis/v1alpha5"
 	"github.com/aws/karpenter-core/pkg/cloudprovider"
-	"github.com/aws/karpenter-core/pkg/utils/machine"
 )
 
 type Drift struct {
-	kubeClient    client.Client
 	cloudProvider cloudprovider.CloudProvider
 }
 
-func (d *Drift) Reconcile(ctx context.Context, provisioner *v1alpha5.Provisioner, node *v1.Node) (reconcile.Result, error) {
-	// If the node is marked as voluntarily disrupted by another controller, do nothing.
-	val, hasAnnotation := node.Annotations[v1alpha5.VoluntaryDisruptionAnnotationKey]
-	if hasAnnotation && val != v1alpha5.VoluntaryDisruptionDriftedAnnotationValue {
+func (d *Drift) Reconcile(ctx context.Context, _ *v1alpha5.Provisioner, machine *v1alpha5.Machine) (reconcile.Result, error) {
+	if !machine.StatusConditions().GetCondition(v1alpha5.MachineLaunched).IsTrue() {
+		return reconcile.Result{}, nil
+	}
+	// If the machine is marked as voluntarily disrupted by another controller, do nothing.
+	voluntarilyDisrupted := machine.StatusConditions().GetCondition(v1alpha5.MachineVoluntarilyDisrupted)
+	if voluntarilyDisrupted.IsTrue() && voluntarilyDisrupted.Reason != v1alpha5.VoluntarilyDisruptedReasonDrifted {
 		return reconcile.Result{}, nil
 	}
 
 	// From here there are three scenarios to handle:
 	// 1. If drift is not enabled but the node is drifted, remove the annotation
 	//    so another disruption controller can annotate the node.
 	if !settings.FromContext(ctx).DriftEnabled {
-		if val == v1alpha5.VoluntaryDisruptionDriftedAnnotationValue {
-			delete(node.Annotations, v1alpha5.VoluntaryDisruptionAnnotationKey)
-			logging.FromContext(ctx).Debugf("removing drift annotation from node as drift has been disabled")
+		if voluntarilyDisrupted.IsTrue() {
+			_ = machine.StatusConditions().ClearCondition(v1alpha5.MachineVoluntarilyDisrupted)
+			logging.FromContext(ctx).Debugf("removing drift status condition from machine as drift has been disabled")
 		}
 		return reconcile.Result{}, nil
 	}
 
-	drifted, err := d.cloudProvider.IsMachineDrifted(ctx, machine.NewFromNode(node))
+	drifted, err := d.cloudProvider.IsMachineDrifted(ctx, machine)
 	if err != nil {
-		return reconcile.Result{}, fmt.Errorf("getting drift for node, %w", err)
+		return reconcile.Result{}, fmt.Errorf("getting drift for machine, %w", err)
 	}
 	// 2. Otherwise, if the node isn't drifted, but has the annotation, remove it.
-	if !drifted && hasAnnotation {
-		delete(node.Annotations, v1alpha5.VoluntaryDisruptionAnnotationKey)
-		logging.FromContext(ctx).Debugf("removing drift annotation from node")
+	if !drifted && voluntarilyDisrupted.IsTrue() {
+		_ = machine.StatusConditions().ClearCondition(v1alpha5.MachineVoluntarilyDisrupted)
+		logging.FromContext(ctx).Debugf("removing drifted status condition from machine")
```

Comment:
Same comment here, it's definitely likely that I don't understand enough about what's going on, but it seems to me that the `MachineVoluntarilyDisrupted` annotation is shared in common by the "drift", "expiration", and "emptiness" reconcilers, either of which is sufficient to mark a machine for disruption. Is my understanding wrong?

- "